cmd/kola/kola.go

```diff
@@ -358,3 +382,169 @@ func runHttpServer(cmd *cobra.Command, args []string) {
 	fmt.Fprintf(os.Stdout, "Serving HTTP on port: %d\n", httpPort)
 	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", httpPort), nil))
 }
+
+func preRunUpgrade(cmd *cobra.Command, args []string) error {
+	// unlike `kola run`, we *require* the --cosa-build arg -- XXX: figure out
+	// how to get this working using cobra's MarkFlagRequired()
+	if kola.Options.CosaBuild == "" {
+		errors.New("Error: missing required argument --cosa-build")
+	}
+
+	err := syncOptions()
+	if err != nil {
+		return err
+	}
+
+	if findParentImage {
+		err = syncFindParentImageOptions()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func postRunUpgrade(cmd *cobra.Command, args []string) {
+	if qemuImageDir != "" && qemuImageDirIsTemp {
+		os.RemoveAll(qemuImageDir)
+	}
+}
+
+// syncFindParentImageOptions handles --find-parent-image automagic.
+func syncFindParentImageOptions() error {
+	var err error
+
+	var parentBaseUrl string
+	switch kola.Options.Distribution {
+	case "fcos":
+		parentBaseUrl, err = getParentFcosBuildBase()
+		if err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("--find-parent-image not yet supported for distro %s", kola.Options.Distribution)
+	}
+
+	var parentCosaBuild *cosa.Build
+	parentCosaBuild, err = cosa.FetchAndParseBuild(parentBaseUrl + "meta.json")
+	if err != nil {
+		return err
+	}
+
+	// Here we handle the --fetch-parent-image --> platform-specific options
+	// based on its cosa build metadata
+	switch kolaPlatform {
+	case "qemu-unpriv":
+		if qemuImageDir == "" {
+			if qemuImageDir, err = ioutil.TempDir("", "kola-run-upgrade"); err != nil {
+				return err
+			}
+			qemuImageDirIsTemp = true
+		}
+		qcowUrl := parentBaseUrl + parentCosaBuild.Images.QEMU.Path
+		qcowLocal := filepath.Join(qemuImageDir, parentCosaBuild.Images.QEMU.Path)
+		decompressedQcowLocal, err := downloadImageAndDecompress(qcowUrl, qcowLocal)
+		if err != nil {
+			return err
+		}
+		kola.QEMUOptions.DiskImage = decompressedQcowLocal
+	case "aws":
+		kola.AWSOptions.AMI, err = parentCosaBuild.FindAMI(kola.AWSOptions.Region)
```

Comment:
See also https://github.com/openshift/installer/issues/2906 - we'll likely at some point need to copy the code from the installer to make images from storage; which gets into terraform vs something else here, or forking out to `openshift-install instantiate-coreos-image` or someting.